test/utils/dirft1d.hpp

```diff
@@ -1,82 +1,124 @@
 
-template<typename T>
-void dirft1d1(BIGINT nj, T *x, std::complex<T> *c, int iflag, BIGINT ms,
-              std::complex<T> *f)
-/* Direct computation of 1D type-1 nonuniform FFT. Interface same as finufft1d1.
-c                  nj-1
-c     f[k1]    =   SUM  c[j] exp(+-i k1 x[j])
-c                  j=0
-c
-c     for -ms/2 <= k1 <= (ms-1)/2.
-c     The output array is in increasing k1 ordering. If iflag>0 the + sign is
-c     used, otherwise the - sign is used, in the exponential.
-*  Uses C++ complex type and winding trick.  Barnett 1/25/17
-*/
-{
-  BIGINT kmin = -(ms / 2);        // integer divide
-  for (BIGINT m = 0; m < ms; ++m)
-    f[m] = std::complex<T>(0, 0); // it knows f is complex type
   for (BIGINT j = 0; j < nj; ++j) {
-    std::complex<T> a  = (iflag > 0) ? exp(std::complex<T>(0, 1) * x[j])
-                                     : exp(-std::complex<T>(0, 1) * x[j]);
-    std::complex<T> p  = pow(a, (T)kmin); // starting phase for most neg freq
-    std::complex<T> cc = c[j];            // no 1/nj prefac
+    FLT xj        = xArr[j];
+    Complex phase = exp(I0 * xj);
+    Complex p     = pow(phase, static_cast<FLT>(kmin));
+    Complex cj    = cArr[j];
     for (BIGINT m = 0; m < ms; ++m) {
-      f[m] += cc * p;
-      p *= a;
+      fArr[m] += cj * p;
+      p *= phase;
     }
   }
 }
 
-template<typename T>
-void dirft1d2(BIGINT nj, T *x, std::complex<T> *c, int iflag, BIGINT ms,
-              std::complex<T> *f)
-/* Direct computation of 1D type-2 nonuniform FFT. Interface same as finufft1d2
-c
-c     c[j] = SUM   f[k1] exp(+-i k1 x[j])
-c             k1
-c                            for j = 0,...,nj-1
-c     where sum is over -ms/2 <= k1 <= (ms-1)/2.
-c     The input array is in increasing k1 ordering. If iflag>0 the + sign is
-c     used, otherwise the - sign is used, in the exponential.
-*  Uses C++ complex type and winding trick.  Barnett 1/25/17
-*/
-{
-  BIGINT kmin = -(ms / 2); // integer divide
+// ------------------------------------------------------------
+// 1D type‑2 NUFFT, direct:
+//   c[j] = Σ_{m=0..ms-1} f[m] exp(i*iflag * k_m * x[j])
+// ------------------------------------------------------------
+template<typename BIGINT,
+         typename XArr, // xArr[j] yields real FLT
+         typename CArr, // cArr[j] yields Complex<FLT>
+         typename FArr  // fArr[m] yields Complex<FLT>
+         >
+void dirft1d2(const BIGINT nj,
+              const XArr &xArr,
+              CArr &cArr,
+              const int iflag,
+              const BIGINT ms,
+              const FArr &fArr) {
+  using Complex = std::decay_t<decltype(cArr[0])>;
+  using FLT     = typename Complex::value_type;
+
+  static_assert(std::is_floating_point<FLT>::value,
+                "dirft1d2: Complex::value_type must be a floating point");
+  static_assert(std::is_same<std::decay_t<decltype(xArr[0])>, FLT>::value,
+                "dirft1d2: xArr element type must match Complex::value_type");
+
+  BIGINT kmin = -(ms / 2);
```

Comment:
I notice you removed the useful comment  "// integer divide" .   I had it there for a reason, since it's a good reminder of the meaning of this. Could you add it back?

Reply:
then I need to make sure that bigint is an int here.